src/mscorlib/src/System/Lazy.cs

```diff
@@ -27,12 +27,158 @@
+            {
+                case LazyState.NoneViaConstructor:
+                case LazyState.NoneViaFactory:
+                case LazyState.NoneException:
+                    return LazyThreadSafetyMode.None;
+
+                case LazyState.PublicationOnlyViaConstructor:
+                case LazyState.PublicationOnlyViaFactory:
+                case LazyState.PublicationOnlyWait:
+                case LazyState.PublicationOnlyException:
+                    return LazyThreadSafetyMode.PublicationOnly;
+
+                case LazyState.ExecutionAndPublicationViaConstructor:
+                case LazyState.ExecutionAndPublicationViaFactory:
+                case LazyState.ExecutionAndPublicationException:
+                    return LazyThreadSafetyMode.ExecutionAndPublication;
+
+                default:
+                    throw InvalidLogic();
+            }
+        }
+
+        internal static LazyThreadSafetyMode GetMode(LazyNonGeneric state)
+        {
+            if (state == null)
+                return LazyThreadSafetyMode.None; // we don't know the mode anymore
+            return state.GetMode();
+        }
+
+        private bool GetIsValueFaulted()
+        {
+            return _exceptionDispatch != null;
+        }
+
+        internal static bool GetIsValueFaulted(LazyNonGeneric state)
+        {
+            if (state == null)
+                return false;
+            return state.GetIsValueFaulted();
+        }
+
+        internal static LazyNonGeneric Create(LazyThreadSafetyMode mode, bool useDefaultConstructor)
+        {
+            if (mode == LazyThreadSafetyMode.None)
+            {
+                return useDefaultConstructor ? NoneViaConstructor : NoneViaFactory;
+            }
+
+            if (mode == LazyThreadSafetyMode.PublicationOnly)
+            {
+                return useDefaultConstructor ? PublicationOnlyViaConstructor : PublicationOnlyViaFactory;
+            }
+
+            if (mode == LazyThreadSafetyMode.ExecutionAndPublication)
+            {
+                // we need to create an object for ExecutionAndPublication because we use Monitor-based locking
+                var state = useDefaultConstructor ? LazyState.ExecutionAndPublicationViaConstructor : LazyState.ExecutionAndPublicationViaFactory;
+                return new LazyNonGeneric(state);
+            }
+
+            throw new ArgumentOutOfRangeException(nameof(mode), Environment.GetResourceString("Lazy_ctor_ModeInvalid"));
+        }
+
+        internal static object CreateViaDefaultConstructor(Type type)
+        {
+            try
+            {
+                return Activator.CreateInstance(type);
+            }
+            catch (MissingMethodException)
+            {
+                throw new MissingMemberException(Environment.GetResourceString("Lazy_CreateValue_NoParameterlessCtorForT"));
+            }
+        }
+
+        internal static LazyThreadSafetyMode GetModeFromIsThreadSafe(bool isThreadSafe)
```

Comment:
This is tiny method that will get inlined into the generic code anyway. You can keep it as it is used to be if you would like.